File: python/kohaku/src/kohaku/lineage/analytics.py
"""Usage analytics (B3. Port of TS packages/lineage/src/analytics.ts).

A pure function that folds the raw lineage event stream into an aggregate summary through which
operators can survey fallback rate, tier distribution, and latency. Read-only, with no I/O and no
clock (easy to test, cache-safe).

Differences from TS:
- LineageSummary / IntentUsage etc. are dataclasses. tiers / cache / promotions / fixations /
  byKind are fixed-key counters, so they stay as dict[str, int] (corresponding to TS plain objects).
- Quantile / rate computation is identical to TS (nearest-rank / total / (composed+total)).
"""

from __future__ import annotations

import math
from collections.abc import Sequence
from dataclasses import dataclass
from typing import Any

from kohaku.spec import LineageEventRecord

TOP_INTENTS_DEFAULT = 10
TOP_INTENTS_MAX = 50


@dataclass(frozen=True)
class IntentUsage:
    """One row of a frequently-used intent (per view.composed intentHash, with canonical alongside)."""

    intentHash: str
    canonical: str
    count: int


@dataclass(frozen=True)
class FallbackSummary:
    total: int
    byKind: dict[str, int]
    """By kind (generation / negotiation / unspecified)."""
    rate: float
    """Fallback rate = total / (composed + total). 0 when the denominator is 0. 0..1."""


@dataclass(frozen=True)
class DurationSummary:
    """Quantiles (milliseconds) of view.composed payload.durationMs. Quantiles are None when the population is 0."""

    count: int
    p50: float | None
    p95: float | None
    p99: float | None
    max: float | None

    def to_wire(self) -> dict[str, Any]:
        """Always spell out the 4 quantile keys (emit `null` even when the population is 0).

        to_jsonable's dataclass default drops None fields, but TS (a plain object) returns
        `p50/p95/p99/max` as explicit `null` when the population is 0 (packages/lineage/src/
        analytics.ts's quantile -> null; JSON.stringify emits each key). durationMs alone needs
        this to_wire so these 4 keys become explicit null (the omission behavior of the other
        dataclasses is unchanged).
        """
        return {
            "count": self.count,
            "p50": self.p50,
            "p95": self.p95,
            "p99": self.p99,
            "max": self.max,
        }


@dataclass(frozen=True)
class LineageSummary:
    """Lineage aggregate summary (the body of the analytics.read response)."""

    events: int
    composed: int
    tiers: dict[str, int]
    cache: dict[str, int]
    fallback: FallbackSummary
    durationMs: DurationSummary
    topIntents: list[IntentUsage]
    promotions: dict[str, int]
    fixations: dict[str, int]


@dataclass(frozen=True)
class SummarizeLineageOptions:
    """Filtering and formatting options for summarize_lineage."""

    tenant: str | None = None
    since: str | None = None
    until: str | None = None
    topIntentsLimit: int | None = None


def _quantile(sorted_values: list[float], p: float) -> float | None:
    """Nearest-rank percentile of an ascending-sorted array (None if empty)."""
    if len(sorted_values) == 0:
        return None
    rank = math.ceil((p / 100) * len(sorted_values))
    idx = min(max(rank - 1, 0), len(sorted_values) - 1)
    return sorted_values[idx]


def summarize_lineage(
    events: Sequence[LineageEventRecord],
    opts: SummarizeLineageOptions | None = None,
) -> LineageSummary:
    """Fold the raw lineage event stream into an aggregate summary (pure function, read-only)."""
    opts = opts if opts is not None else SummarizeLineageOptions()
    scoped: list[LineageEventRecord] = []
    for e in events:
        if opts.tenant is not None and e.tenant != opts.tenant:
            continue
        if opts.since is not None and e.ts < opts.since:
            continue
        if opts.until is not None and e.ts > opts.until:
            continue
        scoped.append(e)

    tiers = {"L0": 0, "L1": 0, "L2": 0}
    cache = {"hit": 0, "miss": 0, "bypass": 0, "fixated": 0, "other": 0}
    by_kind = {"generation": 0, "negotiation": 0, "unspecified": 0}
    promotions = {
        "generated": 0,
        "used": 0,
        "nominated": 0,
        "judged": 0,
        "reviewed": 0,
        "published": 0,
        "withdrawn": 0,
    }
    fixations = {"fixated": 0, "unfixated": 0}
    durations: list[float] = []
    # intentHash -> count (preserving insertion order) + canonical (first seen).
    intent_counts: dict[str, int] = {}
    intent_canonical: dict[str, str] = {}

    composed = 0
    fallback_total = 0

    for e in scoped:
        if e.type == "view.composed":
            composed += 1
            tier = str(e.payload.get("tier") or "")
            if tier in tiers:
                tiers[tier] += 1
            cache_key = str(e.payload.get("cache") or "")
            if cache_key in ("hit", "miss", "bypass", "fixated"):
                cache[cache_key] += 1
            else:
                cache["other"] += 1
            d = e.payload.get("durationMs")
            if isinstance(d, (int, float)) and not isinstance(d, bool) and math.isfinite(d):
                durations.append(d)
            intent_hash = e.payload.get("intentHash")
            if isinstance(intent_hash, str) and len(intent_hash) > 0:
                if intent_hash in intent_counts:
                    intent_counts[intent_hash] += 1
                else:
                    intent_counts[intent_hash] = 1
                    intent_canonical[intent_hash] = str(e.payload.get("canonical") or "")
        elif e.type == "view.fallback":
            fallback_total += 1
            kind = e.payload.get("kind")
            if kind in ("generation", "negotiation"):
                by_kind[kind] += 1
            else:
                by_kind["unspecified"] += 1
        elif e.type == "component.generated":
            promotions["generated"] += 1
        elif e.type == "component.used":
            promotions["used"] += 1
        elif e.type == "component.nominated":
            promotions["nominated"] += 1
        elif e.type == "component.judged":
            promotions["judged"] += 1
        elif e.type == "component.reviewed":
            promotions["reviewed"] += 1
        elif e.type == "component.published":
            promotions["published"] += 1
        elif e.type == "component.withdrawn":
            promotions["withdrawn"] += 1
        elif e.type == "intent.fixated":
            fixations["fixated"] += 1
        elif e.type == "intent.unfixated":
            fixations["unfixated"] += 1

    durations.sort()
    denom = composed + fallback_total
    limit = opts.topIntentsLimit if opts.topIntentsLimit is not None else TOP_INTENTS_DEFAULT
    top_n = min(max(math.floor(limit), 1), TOP_INTENTS_MAX)
    top_intents = sorted(
        (
            IntentUsage(intentHash=h, canonical=intent_canonical[h], count=c)
            for h, c in intent_counts.items()
        ),
        key=lambda u: u.count,
        reverse=True,
    )[:top_n]

    return LineageSummary(
        events=len(scoped),
        composed=composed,
        tiers=tiers,
        cache=cache,
        fallback=FallbackSummary(
            total=fallback_total,
            byKind=by_kind,
            rate=(fallback_total / denom) if denom > 0 else 0,
        ),
        durationMs=DurationSummary(
            count=len(durations),
            p50=_quantile(durations, 50),
            p95=_quantile(durations, 95),
            p99=_quantile(durations, 99),
            max=durations[-1] if len(durations) > 0 else None,
        ),
        topIntents=top_intents,
        promotions=promotions,
        fixations=fixations,
    )

File: packages/lineage/src/analytics.ts
import type { LineageEventRecord } from "@kohaku-ui/spec-core";

/**
 * Usage analytics. A pure function that folds the raw lineage event stream (the LineageEventRecord[]
 * returned by GET /lineage) into an aggregate summary through which operators can survey the fallback rate,
 * tier distribution, and latency.
 *
 * Design approach:
 * - **Read-only**. It never modifies the event schema (the type catalog / payload in events.ts). It aggregates
 *   "as far as it can" from already-recorded payloads (view.composed's tier / cache / durationMs / intentHash /
 *   canonical, view.fallback's kind, etc.). It does not require any new payload field.
 * - Pure function. It aggregates by looking only at the input array, holding no I/O and no clock (easy to test,
 *   cache-safe).
 * - Scope narrowing (tenant / since / until) is also taken as arguments to pre-filter the array. The route side
 *   passes an array already narrowed by the window via storage's listLineage, but this function alone can do the
 *   same narrowing (double application is idempotent).
 */

/** One row for a frequently-used intent (per view.composed intentHash, with canonical noted alongside). */
export interface IntentUsage {
  intentHash: string;
  canonical: string;
  count: number;
}

/** Lineage aggregate summary (the body of the analytics.read response). */
export interface LineageSummary {
  /** Total number of events aggregated (after narrowing by tenant / since / until). */
  events: number;
  /** Count of view.composed (number of successful view presentations). */
  composed: number;
  /** Tier distribution (view.composed's payload.tier). */
  tiers: { L0: number; L1: number; L2: number };
  /**
   * Cache breakdown (view.composed's payload.cache). The 4 known kinds (hit/miss/bypass/fixated) + anything
   * unexpected goes to other. The provenance.cache vocabulary is owned by the composer (this function only
   * counts the known vocabulary and lumps unknown values into other).
   */
  cache: { hit: number; miss: number; bypass: number; fixated: number; other: number };
  /** Fallback (view.fallback). */
  fallback: {
    total: number;
    /** Per kind (generation / negotiation; an unspecified payload.kind counts as unspecified). */
    byKind: { generation: number; negotiation: number; unspecified: number };
    /**
     * Fallback rate = total / (composed + total). "The fraction of view-presentation attempts that dropped to
     * fallback." When the denominator is 0 (no presentations and no fallbacks) it is 0. A real number in 0..1.
     */
    rate: number;
  };
  /**
   * Quantiles of view.composed's payload.durationMs (milliseconds). Only events that carry durationMs as a number
   * form the population (the "aggregate as far as you can" policy; older events without durationMs are excluded
   * from the population). When the population is 0 the quantiles are null. Quantiles are nearest-rank
   * (the ceil(p/100 · n)-th value).
   */
  durationMs: {
    count: number;
    p50: number | null;
    p95: number | null;
    p99: number | null;
    max: number | null;
  };
  /** Top N frequently-used intents (per view.composed intentHash, in descending count order. Default N=10). */
  topIntents: IntentUsage[];
  /** Event counts for the promotion lifecycle (each component.* transition). */
  promotions: {
    generated: number;
    used: number;
    nominated: number;
    /** component.schemaSuggested (a machine proposal was attached at nomination). */
    schemaSuggested: number;
    judged: number;
    reviewed: number;
    /** component.schemaEdited (a reviewer approved a candidate that carried a suggestion; changed may be empty). */
    schemaEdited: number;
    published: number;
    withdrawn: number;
  };
  /**
   * Human review turnaround: the time from a candidate's `component.nominated` to the next `component.reviewed`
   * whose decision is approve or reject, paired per (tenant, artifactId). A requestChanges decision does not
   * complete a review (the candidate is re-nominated later and measured again from that nomination). A
   * reviewed event with no preceding nomination in the window is ignored. Quantiles are nearest-rank like
   * durationMs. `acceptedAsIs` counts component.schemaEdited records whose `changed` is empty (the machine
   * suggestion was approved without any edit) — the ticket's "zero-edit approval" KPI.
   */
  review: {
    count: number;
    durationMs: { p50: number | null; p95: number | null; max: number | null };
    acceptedAsIs: number;
  };
  /** Fixation event counts (intent.fixated / intent.unfixated). */
  fixations: { fixated: number; unfixated: number };
}

/** Narrowing and shaping options for summarizeLineage. */
export interface SummarizeLineageOptions {
  /** Narrow by tenant (aggregate only records whose record.tenant matches). Older events without a recorded tenant are excluded as non-matching. */
  tenant?: string;
  /** Aggregate only events at or after this time (record.ts >= since). Expects canonical ISO8601 (the route normalizes before passing). */
  since?: string;
  /** Aggregate only events at or before this time (record.ts <= until). Expects canonical ISO8601. */
  until?: string;
  /** N for topIntents. Default 10, clamped to 1..50. */
  topIntentsLimit?: number;
}

const TOP_INTENTS_DEFAULT = 10;
const TOP_INTENTS_MAX = 50;

/** Folds the raw lineage event stream into an aggregate summary (pure, read-only). */
export function summarizeLineage(
  events: readonly LineageEventRecord[],
  opts: SummarizeLineageOptions = {},
): LineageSummary {
  const scoped = events.filter((e) => {
    if (opts.tenant != null && e.tenant !== opts.tenant) return false;
    if (opts.since != null && e.ts < opts.since) return false;
    if (opts.until != null && e.ts > opts.until) return false;
    return true;
  });

  const tiers = { L0: 0, L1: 0, L2: 0 };
  const cache = { hit: 0, miss: 0, bypass: 0, fixated: 0, other: 0 };
  const byKind = { generation: 0, negotiation: 0, unspecified: 0 };
  const promotions = {
    generated: 0,
    used: 0,
    nominated: 0,
    schemaSuggested: 0,
    judged: 0,
    reviewed: 0,
    schemaEdited: 0,
    published: 0,
    withdrawn: 0,
  };
  const fixations = { fixated: 0, unfixated: 0 };
  const durations: number[] = [];
  // intentHash → { canonical (first seen), count }. Preserves insertion order while taking the top items by descending count.
  const intents = new Map<string, { canonical: string; count: number }>();

  let composed = 0;
  let fallbackTotal = 0;
  // (tenant, artifactId) -> ts of the latest nomination not yet closed by an approve/reject review.
  const openNominations = new Map<string, string>();
  const reviewDurations: number[] = [];
  let acceptedAsIs = 0;
  const reviewKey = (e: LineageEventRecord): string =>
    `${e.tenant ?? ""}\u0000${String(e.payload["artifactId"] ?? "")}`;

  for (const e of scoped) {
    switch (e.type) {
      case "view.composed": {
        composed++;
        const tier = String(e.payload["tier"] ?? "");
        if (tier === "L0" || tier === "L1" || tier === "L2") tiers[tier]++;
        const cacheKey = String(e.payload["cache"] ?? "");
        if (cacheKey === "hit" || cacheKey === "miss" || cacheKey === "bypass" || cacheKey === "fixated") {
          cache[cacheKey]++;
        } else {
          cache.other++;
        }
        const d = e.payload["durationMs"];
        if (typeof d === "number" && Number.isFinite(d)) durations.push(d);
        const intentHash = e.payload["intentHash"];
        if (typeof intentHash === "string" && intentHash.length > 0) {
          const prev = intents.get(intentHash);
          if (prev != null) prev.count++;
          else intents.set(intentHash, { canonical: String(e.payload["canonical"] ?? ""), count: 1 });
        }
        break;
      }
      case "view.fallback": {
        fallbackTotal++;
        const kind = e.payload["kind"];
        if (kind === "generation" || kind === "negotiation") byKind[kind]++;
        else byKind.unspecified++;
        break;
      }
      case "component.generated":
        promotions.generated++;
        break;
      case "component.used":
        promotions.used++;
        break;
      case "component.nominated":
        promotions.nominated++;
        openNominations.set(reviewKey(e), e.ts);
        break;
      case "component.schemaSuggested":
        promotions.schemaSuggested++;
        break;
      case "component.judged":
        promotions.judged++;
        break;
      case "component.reviewed": {
        promotions.reviewed++;
        const decision = e.payload["decision"];
        if (decision === "approve" || decision === "reject") {
          const key = reviewKey(e);
          const nominatedAt = openNominations.get(key);
          if (nominatedAt != null) {
            const delta = Date.parse(e.ts) - Date.parse(nominatedAt);
            if (Number.isFinite(delta) && delta >= 0) reviewDurations.push(delta);
            openNominations.delete(key);
          }
        }
        break;
      }
      case "component.schemaEdited": {
        promotions.schemaEdited++;
        const changed = e.payload["changed"];
        if (Array.isArray(changed) && changed.length === 0) acceptedAsIs++;
        break;
      }
      case "component.published":
        // Counts a promotion/reconcile audit backfill (payload.reconciled:true) the same as the original
        // synchronous record — the projection was published exactly once either way, so double-counting the
        // *fact* of publication is intentional here (only the audit log entry was delayed, not the publish itself).
        promotions.published++;
        break;
      case "component.withdrawn":
        promotions.withdrawn++;
        break;
      case "intent.fixated":
        fixations.fixated++;
        break;
      case "intent.unfixated":
        fixations.unfixated++;
        break;
      default:
        break;
    }
  }

  durations.sort((a, b) => a - b);
  reviewDurations.sort((a, b) => a - b);
  const denom = composed + fallbackTotal;
  const topN = Math.min(
    Math.max(Math.floor(opts.topIntentsLimit ?? TOP_INTENTS_DEFAULT), 1),
    TOP_INTENTS_MAX,
  );
  const topIntents: IntentUsage[] = [...intents.entries()]
    .map(([intentHash, v]) => ({ intentHash, canonical: v.canonical, count: v.count }))
    .sort((a, b) => b.count - a.count)
    .slice(0, topN);

  return {
    events: scoped.length,
    composed,
    tiers,
    cache,
    fallback: {
      total: fallbackTotal,
      byKind,
      rate: denom > 0 ? fallbackTotal / denom : 0,
    },
    durationMs: {
      count: durations.length,
      p50: quantile(durations, 50),
      p95: quantile(durations, 95),
      p99: quantile(durations, 99),
      max: durations.length > 0 ? durations[durations.length - 1]! : null,
    },
    topIntents,
    promotions,
    review: {
      count: reviewDurations.length,
      durationMs: {
        p50: quantile(reviewDurations, 50),
        p95: quantile(reviewDurations, 95),
        max: reviewDurations.length > 0 ? reviewDurations[reviewDurations.length - 1]! : null,
      },
      acceptedAsIs,
    },
    fixations,
  };
}

/** Nearest-rank percentile of an ascending-sorted array (null if empty). */
function quantile(sorted: number[], p: number): number | null {
  if (sorted.length === 0) return null;
  const rank = Math.ceil((p / 100) * sorted.length);
  const idx = Math.min(Math.max(rank - 1, 0), sorted.length - 1);
  return sorted[idx]!;
}
